op-batcher/batcher/channel_manager.go

```diff
@@ -71,6 +74,14 @@ func (s *channelManager) TxFailed(id txID) {
 	} else {
 		s.log.Warn("unknown transaction marked as failed", "id", id)
 	}
+
+	// If this channel has no submitted transactions, put the pending blocks back into the
+	// local saved blocks and reset this state so it can try to build a new channel.
```

Comment:
What's the purpose of removing the pending channel instead of adding the frames back to it? I think this would cause batches to be slow if the transaction failed

Reply:
This was added to avoid the scenario in which a repeatedly-failing transaction results in an infinite loop during shutdown.

For example, if the batcher does not have enough ETH to cover L1 batch submission, then each call to `TxFailed` will clear the `pendingTransactions` list but keep the failing frame. On the next loop, we will call `TxData` which will detect and return this re-enqueued frame, which could repeat indefinitely.

By clearing the pending channel, this can detect and recover from the scenario where a channel has no submitted frames, and can treat this the same as if no channel was ever constructed in the first place.

It might make the most sense to go with @sebastianst's suggestion below and only applying this behavior when the channel manager is already closed: https://github.com/ethereum-optimism/optimism/pull/5105/files/ba5b58c79604c0821ef28553e6ed01263ae3d544#r1142082400